cfn_update process for EC2 resources

ltsampros opened the pull request:
Partial support for cloudformation updates is now supported on the EC2 and ELB sections of the configuration file:

```fab application:courtfinder aws:my_project_prod environment:dev config:/path/to/courtfinder-dev.yaml tag:[tag_name] cfn_update```

Note: Running this command will show you some structured output of what
changes and how. Also a unified diff is printed on output between the
old and the new Launch Configuration sections. Although we have gone
to great lengths with this command, it can result in destructive
operations, particularly if one reduced the desired/max/min capacities
of the Auto Scaling Group.

Note #2: In order to test this, you will have to set PYTHONPATH pointing to this branch of the PR and to bootstrap-salt's version https://github.com/ministryofjustice/bootstrap-salt/pull/80

Note #3: The fabfile.py of the relevant -deploy repo should contain a line like this: ```from bootstrap_salt.fab_tasks import cfn_create, cfn_delete, cfn_update```

Commit:
cfn_update process for EC2 resources

Commit:
Static boundary on multipart userdata.

Fixes cfn update when no changes are requested. Tthe different
boundary mde CFN think that this was indeed a new LC.

ltsampros commented on bootstrap_cfn/mime_packer.py:
```diff
@@ -46,7 +46,7 @@ def get_type(content, deftype):
 
 def pack(parts, opts={}):
-    outer = MIMEMultipart()
+    outer = MIMEMultipart(boundary='boostrapcfnboundary141')
```
I was thinking of using the library's version here but I think I might revert it to something more static. Versions are possible to change and the scripts to remain the same.

ltsampros commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -687,6 +686,22 @@ def isactive():
+    cfn = get_connection(Cloudformation)
+    # Get online template
+    response = cfn.conn_cfn.get_template(stack_name)
+    body = response['GetTemplateResponse']['GetTemplateResult']['TemplateBody']
```
Some improvements/error checking etc are needed here.

Commit:
Add confirmation and diff output to cfn_update

Commit:
Add unified diff to launch configuration

Commit:
cfn_update documentation

Commit:
Add Update_complete to stack done states

yufangzhang reviewed (commented):
I tested it on adp-dev to resize instances and update ASG and ACM -- part of EC2 and ELB configuration. All changes applied. It may be better if we could highlight the diff changes a bit because currently it's a bit hard to tell.......
Also this code automatically updates docker (new user data etc) so it can break things..

ltsampros commented:
@yufangzhang This code is "updating docker" as a "side-effect". It needs an updated bootstrap-salt and currently the new b-salt version ships with a new urllib3 version. That version is required for new-docker but I don't think it breaks older docker formulas.

Commit:
Various improvements and functionality on cfn_update

- Remove unneeded code, add colors and better logging.
- Cleanup resources in case of stale ELBs and their dependents
- Fix some unhandled cases with cfn_update and a couple of corner cases
- transient DatabaseSG on new object dictionary

Commit:
Better handling of dicts and add 2nd diff method
